### DISCUSSION TITLE
PGN

### DISCUSSION THREAD

**Discussion by EdNekebno** [EdNekebno/DGTCentaurMods#62] [Show and tell]
As everything but lichess generates data now for the database, I've coded in a very basic PGN download bit in the web interface. It's not pretty and it just shows the last 20 games. The idea is that this covers what most people want to do. I can go back at a later point and make it pretty and more feature rich in terms of a game database.

It seems that by default a PGN file has some headers. See the previous paragraph for the user potentially being able to edit them in the future. But this means two things:

1. I have updated the fields in the game table in the sqlite db to have fields for each of these header items. I then emptied the table out and force uploaded the new db to github.
2. I should relook at what extent we can provide this data. e.g. I don't think with eboard or the centaur stockfish that we can really tell which color the player is. But for gamemanager stuff that is maybe a possibility. Also I think for lichess it is too - to fill in the lichess player details, etc.

**Comment by wormstein** [comment]
I'm giving it a try now. Sound a huge leap in features.
Web interface can be impoved with a CSS Framework, I've seen people using Bootstrap to make websites pretty...

For centaur software, you can detect the color by looking in centaur/settings, there is a settings_1_2,dat file that is a dictionary. There is a setting in there called rotated.

**Reply by EdNekebno** [reply]
Cool. Although stockfish needs to detect it...it's hard enough to get stockfish to write the moves to the database...that might take a while longer :)

The PGN implementation is very very basic. I can definitely make everything prettier later.

**Comment by wormstein** [comment]
From my past research on doing something similar, i have some code that figures out the move by comparing two fens... See if this helps you.

``` python
nums = {1:"a", 2:"b", 3:"c", 4:"d", 5:"e", 6:"f", 7:"g", 8:"h"}
def get_uci(board1, board2, who_moved):
    str_board = str(board1).split("\n")
    str_board2 = str(board2).split("\n")
    move = ""
    flip = False
    if who_moved == "w":
        for i in range(8)[::-1]:
            for x in range(15)[::-1]:
                if str_board[i][x] != str_board2[i][x]:
                    if str_board[i][x] == "." and move == "":
                        flip = True
                    move+=str(nums.get(round(x/2)+1))+str(9-(i+1))
    else:
        for i in range(8):
            for x in range(15):
                if str_board[i][x] != str_board2[i][x]:
                    if str_board[i][x] == "." and move == "":
                        flip = True
                    move+=str(nums.get(round(x/2)+1))+str(9-(i+1))
    if flip:
        move = move[2]+move[3]+move[0]+move[1]
    return move
```

Plan was to grab fen of stockfish from fen.log and compare...

**Reply by EdNekebno** [reply]
Looks good. I think I've got moves.
Pulling the player out of the centaur settings file is a problem - as I'd have to do it in c++ from stockfish. Using gamemanager now you get something like the following:

```
[Event "1000 ELO"]
[Site ""]
[Date "2021-10-22 22:12:55"]
[Round ""]
[White "Player"]
[Black "CT800"]
[Result "1-0"]
[Source "ct800.py"]

1. d4 e6 2. Bf4 Nc6 3. e3 Be7 4. c3 Bh4 5. Bd3 d6 6. Nf3 e5 7. dxe5 Qe7 8. exd6
Bxf2+ 9. Kxf2 Qe6 10. dxc7 Qxa2 11. Rxa2 Bh3 12. gxh3 Rd8 13. cxd8=Q+ Nxd8 14.
Qa4+ Nc6 15. Be4 b5 16. Qxb5 g6 17. Qxc6+ Ke7 18. Qd6+ Ke8 19. Bc6# 1-0
```

Just not from the centaur process - you just get the date and the moves. Takebacks are also a problem with centaur. That's maybe liveable for now as I can probably better concentrate on other stuff. Most of the time people will take a PGN almost straight away and dump it in some analysis or another.

Also I can't work out what I'm doing with git/github and this centaur.db database....I keep messing it up! I swear I sorted it, had it in gitignore, but it just really really wants to upload my test data on here!

**Reply by wormstein** [reply]
The only solution to keep db clean is to write a python file to create the db file, which I could call it on install. Put that in db folder and name it createdb.py.

Then, just `git rm --cache centaur.db` and because it's ignored it should not bother you anymore

Alright man, have a good night :))
